osu.Game/Screens/Play/GameplayClockContainer.cs

```diff
@@ -36,92 +38,94 @@ public class GameplayClockContainer : Container, IAdjustableClock, IGameplayCloc
+        protected readonly FramedBeatmapClock GameplayClock;
+
+        protected override Container<Drawable> Content { get; } = new Container { RelativeSizeAxes = Axes.Both };
 
         /// <summary>
         /// Creates a new <see cref="GameplayClockContainer"/>.
         /// </summary>
         /// <param name="sourceClock">The source <see cref="IClock"/> used for timing.</param>
-        public GameplayClockContainer(IClock sourceClock)
+        /// <param name="applyOffsets">Whether to apply platform, user and beatmap offsets to the mix.</param>
+        public GameplayClockContainer(IClock sourceClock, bool applyOffsets = false)
         {
             SourceClock = sourceClock;
 
             RelativeSizeAxes = Axes.Both;
 
-            decoupledClock = new DecoupleableInterpolatingFramedClock { IsCoupled = false };
-            IsPaused.BindValueChanged(OnIsPausedChanged);
-
-            // this will be replaced during load, but non-null for tests which don't add this component to the hierarchy.
-            FramedClock = new FramedClock();
-        }
-
-        protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnlyDependencyContainer parent)
-        {
-            var dependencies = new DependencyContainer(base.CreateChildDependencies(parent));
-
-            FramedClock = CreateGameplayClock(decoupledClock);
-
-            dependencies.CacheAs<IGameplayClock>(this);
+            InternalChildren = new Drawable[]
+            {
+                GameplayClock = new FramedBeatmapClock(sourceClock, applyOffsets) { IsCoupled = false },
+                Content
+            };
 
-            return dependencies;
+            IsPaused.BindValueChanged(OnIsPausedChanged);
         }
 
         /// <summary>
-        /// Starts gameplay.
+        /// Starts gameplay and marks un-paused state.
         /// </summary>
         public virtual void Start()
         {
             ensureSourceClockSet();
 
-            if (!decoupledClock.IsRunning)
+            isPaused.Value = false;
+
+            // the clock may be stopped via internal means (ie. not via `IsPaused`).
+            if (!GameplayClock.IsRunning)
             {
                 // Seeking the decoupled clock to its current time ensures that its source clock will be seeked to the same time
                 // This accounts for the clock source potentially taking time to enter a completely stopped state
-                Seek(FramedClock.CurrentTime);
+                Seek(GameplayClock.CurrentTime);
 
-                decoupledClock.Start();
+                // Delay the start operation to ensure all children components get the initial seek time.
+                // Without this, children may get a start time beyond StartTime without seeing the time has elapsed.
+                // This can manifest as events which should be fired at the precise StartTime not firing.
+                SchedulerAfterChildren.Add(GameplayClock.Start);
```

Comment:
So as far as I've debugged the issue myself, I understand the issue here to two-part:

1. `GameplayClock.Start()` processes the clock after starting the source.
2. `GameplayClock.Update()` which runs before the content is updated, also processes the clock.

Since there is a non-zero time between the `Start()` call and _either_ of the two `ProcessFrame()` calls, by the time children are updated the current time will _always_ be non-zero. Some amount of time will have elapsed even if that amount is 0.0001ms.

`FrameStabilityContainer` immediately takes on this current time for the first frame and always lists zero elapsed time for this frame. And then of particular importance, samples care about if their actuation point occurred somewhere between the current frame and the last frame, based on elapsed time, which is zero, so there's no "last frame", so there was no actuation.

---

I think this comment needs to explain it a bit better. In particular how "children may get a start time beyond StartTime" - which is those `ProcessFrame()` calls.

Reply:
Yeah, that's a pretty good summary. I've rewritten the inline comment, see if it reads better.